src/AbstractActionHandler.ts

```diff
@@ -189,22 +197,64 @@ export abstract class AbstractActionHandler {
    */
   protected async handleActions(
     state: any,
-    block: Block,
     context: any,
+    nextBlock: NextBlock,
     isReplay: boolean,
   ): Promise<void> {
+    const { block } = nextBlock
     const { blockInfo } = block
 
-    const versionedActions = await this.applyUpdaters(state, block, isReplay, context)
+    const versionedActions = await this.applyUpdaters(state, block, context, isReplay)
     if (!isReplay) {
-      this.runEffects(versionedActions, block, context)
+      this.runEffects(versionedActions, context, nextBlock)
     }
 
     await this.updateIndexState(state, block, isReplay, this.handlerVersionName, context)
     this.lastProcessedBlockNumber = blockInfo.blockNumber
     this.lastProcessedBlockHash = blockInfo.blockHash
   }
 
+  private range(start: number, end: number) {
+    return Array(end - start).fill(0).map((_, i: number) => i + start)
+  }
+
+  private runOrDeferEffect(
+    effect: Effect,
+    payload: any,
+    nextBlock: NextBlock,
+    context: any,
+  ) {
+    const { block, lastIrreversibleBlockNumber } = nextBlock
+    if (!effect.deferUntilIrreversible || block.blockInfo.blockNumber <= lastIrreversibleBlockNumber) {
```

Comment:
not required but putting the contents of this into a method would make it very obvious as to what it decides

shouldExecuteNow(effect.deferUntilIrreversible, block.blockInfo.blockNumber, lastIrreversibleBlockNumber)

Reply:
You're suggesting to wrap the function call in another function call? Wouldn't a comment be better?

Reply:
Option one, add a comment to explain the code.

Option two, have the code explain itself.

Reply:
A third option would be a variable.
```
const shouldExecute = !effect.deferUntilIrreversible ...
if (shouldExecute) {
...
}
```

Reply:
The reason being that it makes parsing the if statement easier. I can skip the definition of `shouldExecute` if I want to and just understand the "how" of control flow without understanding the "why"